src/test/resources/processExecution/simulator.ts

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for details.
+
+import * as assert from "assert";
+import * as child_process from "child_process";
+import * as Q from "q";
+
+import {ISpawnResult, ChildProcess} from "../../../common/node/childProcess";
+import {PromiseUtil} from "../../../common/node/promise";
+
+import {IStdOutEvent, IStdErrEvent, IErrorEvent, IExitEvent, ICustomEvent} from "./recording";
+import * as recording from "./recording";
+import * as simulators from "../simulators/childProcess";
+
+export type IEventArguments = recording.IEventArguments;
+export type Recording = recording.Recording;
+
+export interface ISimulationResult {
+    simulatedProcess: child_process.ChildProcess;
+    simulationEnded: Q.Promise<void> | void;
+}
+
+/* The side effects definition has rule to identify when an event with side effects happened in the simulation,
+   and the callback that must be called for the simulator to simulate that side-effect during the tests.
+   e.g.: When the 'projectWasCreated' event happens, we call a callback to actually create the project */
+export interface ISideEffectsDefinition {
+    beforeStart: () => Q.Promise<void>;
+    outputBased: IOutputBasedSideEffectDefinition[];
+    beforeSuccess: (stdout: string, stderr: string) => Q.Promise<void>;
+}
+
+type IOutputBasedSideEffectDefinition = IOutputSingleEventBasedSideEffectDefinition | IWholeOutputBasedSideEffectDefinition;
+
+// Side effects based on analyzing each stdout event individually
+export interface IOutputSingleEventBasedSideEffectDefinition {
+    eventPattern: RegExp;
+    action: () => Q.Promise<void>;
+}
+
+// Side effects based on analyzing the whole stdout of the recording
+export interface IWholeOutputBasedSideEffectDefinition {
```

Comment:
These 2 interfaces are pretty much the same thing... I would be inclined to have only 1 interface, and modify the constructor of `Simulator` to take 2 separate lists (one for whole output, one for single event). I think it would be much clearer that way, where the consumers would explicitly separate the 2